File: preinstall/sbr-config/sbr_config/output.py
"""Colored terminal output, formatting, tables, and interactive prompts."""

import select
import sys
import time
from typing import List, Optional

from .models import PlannedChange, ValidationResult


class Colors:
    """ANSI color codes."""
    RESET = "\033[0m"
    BOLD = "\033[1m"
    DIM = "\033[2m"
    RED = "\033[31m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    BLUE = "\033[34m"
    MAGENTA = "\033[35m"
    CYAN = "\033[36m"
    WHITE = "\033[37m"


class Output:
    """Handles all user-facing terminal output."""

    def __init__(self, color: bool = True, quiet: bool = False):
        self._color = color
        self._quiet = quiet

    def _c(self, code: str, text: str) -> str:
        """Apply color code if colors are enabled."""
        if not self._color:
            return text
        return f"{code}{text}{Colors.RESET}"

    def header(self, text: str) -> None:
        """Print a section header."""
        if self._quiet:
            return
        print()
        print(self._c(Colors.BOLD + Colors.BLUE, text))
        print(self._c(Colors.BOLD + Colors.BLUE, "=" * len(text)))

    def subheader(self, text: str) -> None:
        """Print a sub-section header."""
        if self._quiet:
            return
        print()
        print(self._c(Colors.BOLD + Colors.CYAN, text))
        print(self._c(Colors.DIM, "-" * len(text)))

    def success(self, text: str) -> None:
        """Print a success message."""
        if self._quiet:
            return
        prefix = self._c(Colors.GREEN, "[PASS]")
        print(f"  {prefix} {text}")

    def fail(self, text: str) -> None:
        """Print a failure message."""
        prefix = self._c(Colors.RED, "[FAIL]")
        print(f"  {prefix} {text}")

    def warning(self, text: str) -> None:
        """Print a warning message."""
        prefix = self._c(Colors.YELLOW, "[WARN]")
        print(f"  {prefix} {text}")

    def error(self, text: str) -> None:
        """Print an error message to stderr."""
        prefix = self._c(Colors.RED + Colors.BOLD, "ERROR:")
        print(f"{prefix} {text}", file=sys.stderr)

    def info(self, text: str) -> None:
        """Print an informational message."""
        if self._quiet:
            return
        prefix = self._c(Colors.CYAN, "[INFO]")
        print(f"  {prefix} {text}")

    def dim(self, text: str) -> None:
        """Print dimmed/secondary text."""
        if self._quiet:
            return
        print(f"  {self._c(Colors.DIM, text)}")

    def nl(self) -> None:
        """Print a blank line."""
        if not self._quiet:
            print()

    def banner(self) -> None:
        """Print the tool banner."""
        if self._quiet:
            return
        title = "sbr-config: Source-Based Routing Configurator"
        print()
        print(self._c(Colors.BOLD + Colors.WHITE, title))
        print(self._c(Colors.DIM, "=" * len(title)))

    def interface_table(self, interfaces: list) -> None:
        """Print a formatted table of detected interfaces."""
        if self._quiet:
            return
        # Header
        fmt = "  {:<12} {:<20} {:<16} {}"
        print(fmt.format("INTERFACE", "IP/PREFIX", "GATEWAY", "STATUS"))
        print(fmt.format("-" * 11, "-" * 19, "-" * 15, "-" * 20))
        for iface in interfaces:
            if iface.is_loopback:
                continue
            gw = iface.gateway or "(none)"
            status_parts = []
            if iface.is_default_route_interface:
                status_parts.append(self._c(Colors.GREEN, "DEFAULT ROUTE"))
            if not iface.is_up:
                status_parts.append(self._c(Colors.YELLOW, "DOWN"))
            status = " | ".join(status_parts) if status_parts else self._c(Colors.DIM, "secondary")
            print(fmt.format(iface.name, iface.cidr, gw, status))

    def validation_report(self, results: List[ValidationResult]) -> None:
        """Print a formatted validation report."""
        if not results:
            self.info("No validation checks to report.")
            return

        current_iface = None
        for r in results:
            if r.interface_name != current_iface:
                current_iface = r.interface_name
                self.subheader(f"Interface: {current_iface}")

            if r.is_correct:
                self.success(f"{r.check_name}: {r.current_value}")
            else:
                self.fail(f"{r.check_name}")
                self.dim(f"  Current:  {r.current_value}")
                self.dim(f"  Expected: {r.expected_value}")
                if r.fix_description:
                    self.dim(f"  Fix:      {r.fix_description}")

    def changes_report(self, changes: List[PlannedChange]) -> None:
        """Print proposed changes with explanations."""
        if not changes:
            self.info("No changes needed -- system is correctly configured.")
            return

        for i, change in enumerate(changes, 1):
            tag = self._c(Colors.BOLD + Colors.YELLOW, f"  {i}.")
            desc = self._c(Colors.BOLD, change.description)
            print(f"{tag} {desc}")

            # Command
            cmd_prefix = self._c(Colors.DIM, "     CMD:")
            cmd = self._c(Colors.CYAN, change.command)
            print(f"{cmd_prefix} {cmd}")

            # Reason (word-wrap at ~70 chars with indent)
            reason_prefix = self._c(Colors.DIM, "     WHY:")
            lines = _wrap_text(change.reason, width=65)
            print(f"{reason_prefix} {lines[0]}")
            for line in lines[1:]:
                print(f"          {line}")
            print()

    def prompt_yn(self, question: str, default: bool = False) -> bool:
        """Prompt user for yes/no answer.

        Args:
            question: The question to ask.
            default: Default answer if user just hits Enter.

        Returns:
            True for yes, False for no.
        """
        suffix = "[Y/n]" if default else "[y/N]"
        try:
            answer = input(f"\n{question} {suffix}: ").strip().lower()
        except (EOFError, KeyboardInterrupt):
            print()
            return False

        if not answer:
            return default
        return answer in ("y", "yes")

    def prompt_timed_confirm(self, timeout: int) -> bool:
        """Post-apply dead man's switch confirmation.

        Displays a prominent warning and waits for the user to type 'yes'
        within ``timeout`` seconds.  If no confirmation is received the
        caller should automatically roll back.

        Args:
            timeout: Seconds to wait.  Must be > 0.

        Returns:
            True if the user confirmed, False if timeout expired or input
            was anything other than 'yes'.
        """
        box_w = 66
        border = self._c(Colors.RED + Colors.BOLD, "#" * box_w)
        pad = self._c(Colors.RED + Colors.BOLD, "#") + " " * (box_w - 2) + self._c(Colors.RED + Colors.BOLD, "#")

        print()
        print(border)
        print(pad)
        self._box_line("CONNECTIVITY CHECK", box_w, Colors.RED + Colors.BOLD)
        print(pad)
        self._box_line("Changes have been applied. Please verify that you", box_w, Colors.YELLOW)
        self._box_line("still have connectivity to this system.", box_w, Colors.YELLOW)
        print(pad)
        self._box_line("Type 'yes' within %d seconds to KEEP the changes." % timeout, box_w, Colors.WHITE + Colors.BOLD)
        self._box_line("If no response is received, all changes will be", box_w, Colors.WHITE)
        self._box_line("AUTOMATICALLY ROLLED BACK for safety.", box_w, Colors.WHITE)
        print(pad)
        print(border)
        print()

        deadline = time.time() + timeout

        # If stdin isn't a TTY (e.g. piped input in validation harness), fall
        # back to line-buffered input with a static prompt -- we can't do the
        # cbreak-mode live redraw without a terminal.
        if not sys.stdin.isatty():
            return self._prompt_timed_confirm_nontty(deadline, timeout)

        return self._prompt_timed_confirm_tty(deadline, timeout)

    def _prompt_timed_confirm_nontty(self, deadline: float, timeout: int) -> bool:
        """Fallback for non-TTY stdin: static prompt, single readline."""
        sys.stdout.write(
            self._c(Colors.YELLOW + Colors.BOLD, "  [%ds]" % timeout)
            + " Type 'yes' to confirm (then press Enter): "
        )
        sys.stdout.flush()
        try:
            remaining = deadline - time.time()
            if remaining <= 0:
                return False
            ready, _, _ = select.select([sys.stdin], [], [], remaining)
        except (OSError, ValueError):
            return False
        if not ready:
            return False
        try:
            answer = sys.stdin.readline().strip().lower()
        except (EOFError, IOError):
            return False
        sys.stdout.write("\n")
        sys.stdout.flush()
        return answer == "yes"

    def _prompt_timed_confirm_tty(self, deadline: float, timeout: int) -> bool:
        """TTY path: put terminal in cbreak mode so we own the echo, then
        redraw the prompt line every second with the live countdown and
        whatever the user has typed so far.  Because we're the one echoing
        keystrokes, we can safely clear and rewrite the line without losing
        the user's in-progress input.
        """
        import termios
        import tty

        fd = sys.stdin.fileno()
        old_settings = termios.tcgetattr(fd)

        typed = ""
        confirmed = False

        def _draw(secs_left):
            # \r -> col 0, \033[K -> erase to end of line
            sys.stdout.write(
                "\r\033[K"
                + self._c(Colors.YELLOW + Colors.BOLD, "  [%2ds]" % int(secs_left + 0.5))
                + " Type 'yes' to confirm: "
                + typed
            )
            sys.stdout.flush()

        try:
            tty.setcbreak(fd)
            _draw(timeout)
            last_secs = int(timeout + 0.5)

            while True:
                remaining = deadline - time.time()
                if remaining <= 0:
                    break

                # Tick the countdown if the displayed second changed.
                cur_secs = int(remaining + 0.5)
                if cur_secs != last_secs:
                    _draw(remaining)
                    last_secs = cur_secs

                # Wake ~10x/sec to keep the countdown smooth and responsive.
                try:
                    ready, _, _ = select.select(
                        [sys.stdin], [], [], min(0.1, remaining)
                    )
                except (OSError, ValueError):
                    break
                if not ready:
                    continue

                try:
                    ch = sys.stdin.read(1)
                except (OSError, IOError):
                    break
                if not ch:
                    break  # EOF

                if ch in ("\n", "\r"):
                    # Enter -- finalize this line
                    sys.stdout.write("\n")
                    sys.stdout.flush()
                    answer = typed.strip().lower()
                    if answer == "yes":
                        confirmed = True
                        break
                    if answer:
                        sys.stdout.write(
                            "  "
                            + self._c(Colors.RED, "Please type exactly 'yes' to confirm.")
                            + "\n"
                        )
                    typed = ""
                    _draw(deadline - time.time())
                    last_secs = int(deadline - time.time() + 0.5)
                elif ch in ("\x7f", "\b"):
                    if typed:
                        typed = typed[:-1]
                        _draw(deadline - time.time())
                elif ch == "\x03":
                    # Ctrl-C -- bail out, let caller roll back
                    break
                elif ch.isprintable():
                    typed += ch
                    # Echo without the full redraw -- cheaper and avoids flicker.
                    sys.stdout.write(ch)
                    sys.stdout.flush()
                # Ignore other control chars / escape sequences
        finally:
            termios.tcsetattr(fd, termios.TCSANOW, old_settings)
            sys.stdout.write("\n")
            sys.stdout.flush()

        return confirmed

    def _box_line(self, text, width, color_code):
        """Print a centered line inside a bordered box."""
        inner = width - 4  # account for "# " and " #"
        padded = text.center(inner)
        left = self._c(Colors.RED + Colors.BOLD, "# ")
        right = self._c(Colors.RED + Colors.BOLD, " #")
        middle = self._c(color_code, padded)
        print(left + middle + right)

    def summary(self, passed: int, failed: int) -> None:
        """Print a validation summary line."""
        total = passed + failed
        if failed == 0:
            msg = self._c(Colors.GREEN + Colors.BOLD, f"All {total} checks passed")
        else:
            msg = (
                self._c(Colors.GREEN, f"{passed} passed") +
                ", " +
                self._c(Colors.RED + Colors.BOLD, f"{failed} failed") +
                f" (out of {total})"
            )
        print()
        print(f"  Summary: {msg}")


def _wrap_text(text: str, width: int = 70) -> List[str]:
    """Simple word-wrap without importing textwrap."""
    words = text.split()
    lines = []
    current_line = ""
    for word in words:
        if current_line and len(current_line) + 1 + len(word) > width:
            lines.append(current_line)
            current_line = word
        else:
            current_line = f"{current_line} {word}" if current_line else word
    if current_line:
        lines.append(current_line)
    return lines or [""]

File: preinstall/sbr-config/internal/output/output.go
// Package output provides colored terminal output, formatted tables,
// and interactive prompts including the dead man's switch.
package output

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/weka/tools/preinstall/sbr-config/internal/models"
)

// ANSI color codes.
const (
	Reset   = "\033[0m"
	Bold    = "\033[1m"
	Dim     = "\033[2m"
	Red     = "\033[31m"
	Green   = "\033[32m"
	Yellow  = "\033[33m"
	Blue    = "\033[34m"
	Magenta = "\033[35m"
	Cyan    = "\033[36m"
	White   = "\033[37m"
)

// Output handles all user-facing terminal output.
type Output struct {
	color bool
	quiet bool
}

// New creates an Output instance.
func New(color, quiet bool) *Output {
	return &Output{color: color, quiet: quiet}
}

// c applies an ANSI color code if colors are enabled.
func (o *Output) c(code, text string) string {
	if !o.color {
		return text
	}
	return code + text + Reset
}

// Header prints a section header.
func (o *Output) Header(text string) {
	if o.quiet {
		return
	}
	fmt.Println()
	fmt.Println(o.c(Bold+Blue, text))
	fmt.Println(o.c(Bold+Blue, strings.Repeat("=", len(text))))
}

// Subheader prints a sub-section header.
func (o *Output) Subheader(text string) {
	if o.quiet {
		return
	}
	fmt.Println()
	fmt.Println(o.c(Bold+Cyan, text))
	fmt.Println(o.c(Dim, strings.Repeat("-", len(text))))
}

// Success prints a success message.
func (o *Output) Success(text string) {
	if o.quiet {
		return
	}
	fmt.Printf("  %s %s\n", o.c(Green, "[PASS]"), text)
}

// Fail prints a failure message.
func (o *Output) Fail(text string) {
	fmt.Printf("  %s %s\n", o.c(Red, "[FAIL]"), text)
}

// Warning prints a warning message.
func (o *Output) Warning(text string) {
	fmt.Printf("  %s %s\n", o.c(Yellow, "[WARN]"), text)
}

// Error prints an error message to stderr.
func (o *Output) Error(text string) {
	fmt.Fprintf(os.Stderr, "%s %s\n", o.c(Red+Bold, "ERROR:"), text)
}

// Info prints an informational message.
func (o *Output) Info(text string) {
	if o.quiet {
		return
	}
	fmt.Printf("  %s %s\n", o.c(Cyan, "[INFO]"), text)
}

// Dim prints dimmed/secondary text.
func (o *Output) Dim(text string) {
	if o.quiet {
		return
	}
	fmt.Printf("  %s\n", o.c(Dim, text))
}

// NL prints a blank line.
func (o *Output) NL() {
	if !o.quiet {
		fmt.Println()
	}
}

// Banner prints the tool banner.
func (o *Output) Banner() {
	if o.quiet {
		return
	}
	title := "sbr-config: Source-Based Routing Configurator"
	fmt.Println()
	fmt.Println(o.c(Bold+White, title))
	fmt.Println(o.c(Dim, strings.Repeat("=", len(title))))
}

// InterfaceTable prints a formatted table of detected interfaces.
func (o *Output) InterfaceTable(interfaces []models.InterfaceInfo) {
	if o.quiet {
		return
	}
	const fmtStr = "  %-12s %-20s %-16s %s\n"
	fmt.Printf(fmtStr, "INTERFACE", "IP/PREFIX", "GATEWAY", "STATUS")
	fmt.Printf(fmtStr,
		strings.Repeat("-", 11),
		strings.Repeat("-", 19),
		strings.Repeat("-", 15),
		strings.Repeat("-", 20),
	)
	for _, iface := range interfaces {
		if iface.IsLoopback {
			continue
		}
		gw := iface.GatewayStr()
		var statusParts []string
		if iface.IsDefaultRouteInterface {
			statusParts = append(statusParts, o.c(Green, "DEFAULT ROUTE"))
		}
		if !iface.IsUp {
			statusParts = append(statusParts, o.c(Yellow, "DOWN"))
		}
		status := strings.Join(statusParts, " | ")
		if status == "" {
			status = o.c(Dim, "secondary")
		}
		fmt.Printf(fmtStr, iface.Name, iface.CIDR(), gw, status)
	}
}

// ValidationReport prints a formatted validation report.
func (o *Output) ValidationReport(results []models.ValidationResult) {
	if len(results) == 0 {
		o.Info("No validation checks to report.")
		return
	}

	currentIface := ""
	for _, r := range results {
		if r.InterfaceName != currentIface {
			currentIface = r.InterfaceName
			o.Subheader(fmt.Sprintf("Interface: %s", currentIface))
		}

		if r.IsCorrect {
			o.Success(fmt.Sprintf("%s: %s", r.CheckName, r.CurrentValue))
		} else {
			o.Fail(r.CheckName)
			o.Dim(fmt.Sprintf("  Current:  %s", r.CurrentValue))
			o.Dim(fmt.Sprintf("  Expected: %s", r.ExpectedValue))
			if r.FixDescription != "" {
				o.Dim(fmt.Sprintf("  Fix:      %s", r.FixDescription))
			}
		}
	}
}

// ChangesReport prints proposed changes with explanations.
func (o *Output) ChangesReport(changes []models.PlannedChange) {
	if len(changes) == 0 {
		o.Info("No changes needed -- system is correctly configured.")
		return
	}

	for i, change := range changes {
		tag := o.c(Bold+Yellow, fmt.Sprintf("  %d.", i+1))
		desc := o.c(Bold, change.Description)
		fmt.Printf("%s %s\n", tag, desc)

		cmdPrefix := o.c(Dim, "     CMD:")
		cmd := o.c(Cyan, change.Command)
		fmt.Printf("%s %s\n", cmdPrefix, cmd)

		reasonPrefix := o.c(Dim, "     WHY:")
		lines := wrapText(change.Reason, 65)
		fmt.Printf("%s %s\n", reasonPrefix, lines[0])
		for _, line := range lines[1:] {
			fmt.Printf("          %s\n", line)
		}
		fmt.Println()
	}
}

// PromptYN prompts the user for a yes/no answer.
func (o *Output) PromptYN(question string, defaultYes bool) bool {
	suffix := "[y/N]"
	if defaultYes {
		suffix = "[Y/n]"
	}
	fmt.Printf("\n%s %s: ", question, suffix)

	scanner := bufio.NewScanner(os.Stdin)
	if !scanner.Scan() {
		fmt.Println()
		return false
	}
	answer := strings.TrimSpace(strings.ToLower(scanner.Text()))
	if answer == "" {
		return defaultYes
	}
	return answer == "y" || answer == "yes"
}

// PromptTimedConfirm implements the post-apply dead man's switch.
// It displays a prominent warning and waits for the user to type 'yes'
// within timeout seconds. Returns true if confirmed, false if timeout
// expired or input was anything other than 'yes'.
func (o *Output) PromptTimedConfirm(timeout int) bool {
	boxW := 66
	border := o.c(Red+Bold, strings.Repeat("#", boxW))
	pad := o.c(Red+Bold, "#") + strings.Repeat(" ", boxW-2) + o.c(Red+Bold, "#")

	fmt.Println()
	fmt.Println(border)
	fmt.Println(pad)
	o.boxLine("CONNECTIVITY CHECK", boxW, Red+Bold)
	fmt.Println(pad)
	o.boxLine("Changes have been applied. Please verify that you", boxW, Yellow)
	o.boxLine("still have connectivity to this system.", boxW, Yellow)
	fmt.Println(pad)
	o.boxLine(fmt.Sprintf("Type 'yes' within %d seconds to KEEP the changes.", timeout), boxW, White+Bold)
	o.boxLine("If no response is received, all changes will be", boxW, White)
	o.boxLine("AUTOMATICALLY ROLLED BACK for safety.", boxW, White)
	fmt.Println(pad)
	fmt.Println(border)
	fmt.Println()

	deadline := time.Now().Add(time.Duration(timeout) * time.Second)

	// Channel to receive stdin lines from a goroutine.
	lineCh := make(chan string, 1)
	go func() {
		scanner := bufio.NewScanner(os.Stdin)
		for scanner.Scan() {
			lineCh <- scanner.Text()
		}
		close(lineCh)
	}()

	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		remaining := time.Until(deadline)
		if remaining <= 0 {
			break
		}

		secs := int(remaining.Seconds() + 0.5)
		fmt.Fprintf(os.Stdout, "\r%s Type 'yes' to confirm: ",
			o.c(Yellow+Bold, fmt.Sprintf("  [%2ds]", secs)))

		select {
		case line, ok := <-lineCh:
			if !ok {
				// stdin closed
				goto done
			}
			answer := strings.TrimSpace(strings.ToLower(line))
			if answer == "yes" {
				fmt.Fprintf(os.Stdout, "\r%s\r", strings.Repeat(" ", 60))
				return true
			}
			if answer != "" {
				fmt.Fprintf(os.Stdout, "  %s\n", o.c(Red, "Please type exactly 'yes' to confirm."))
			}
		case <-ticker.C:
			// Redraw countdown
		}
	}

done:
	fmt.Fprintf(os.Stdout, "\r%s\r", strings.Repeat(" ", 60))
	return false
}

// boxLine prints a centered line inside a bordered box.
func (o *Output) boxLine(text string, width int, colorCode string) {
	inner := width - 4 // account for "# " and " #"
	padded := centerStr(text, inner)
	left := o.c(Red+Bold, "# ")
	right := o.c(Red+Bold, " #")
	middle := o.c(colorCode, padded)
	fmt.Println(left + middle + right)
}

// Summary prints a validation summary line.
func (o *Output) Summary(passed, failed int) {
	total := passed + failed
	var msg string
	if failed == 0 {
		msg = o.c(Green+Bold, fmt.Sprintf("All %d checks passed", total))
	} else {
		msg = o.c(Green, fmt.Sprintf("%d passed", passed)) +
			", " +
			o.c(Red+Bold, fmt.Sprintf("%d failed", failed)) +
			fmt.Sprintf(" (out of %d)", total)
	}
	fmt.Println()
	fmt.Printf("  Summary: %s\n", msg)
}

// wrapText is a simple word-wrap without importing textwrap.
func wrapText(text string, width int) []string {
	words := strings.Fields(text)
	if len(words) == 0 {
		return []string{""}
	}
	var lines []string
	current := ""
	for _, word := range words {
		if current != "" && len(current)+1+len(word) > width {
			lines = append(lines, current)
			current = word
		} else if current != "" {
			current += " " + word
		} else {
			current = word
		}
	}
	if current != "" {
		lines = append(lines, current)
	}
	return lines
}

// centerStr center-pads a string within the given width.
func centerStr(s string, width int) string {
	if len(s) >= width {
		return s
	}
	left := (width - len(s)) / 2
	right := width - len(s) - left
	return strings.Repeat(" ", left) + s + strings.Repeat(" ", right)
}
